models/template.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+
+const Joi = require('joi');
+const mutate = require('../lib/mutate');
+const Template = require('../config/template');
+const scmUri = Joi.reach(require('./pipeline').base, 'scmUri');
+
+const MODEL = {
+    id: Joi
+        .number().integer().positive()
+        .description('Identifier of this template')
+        .example(123345),
+
+    labels: Joi
+        .array()
+        .items(Joi.string())
+        .description('Labels for template')
+        .example(['stable', 'latest', 'beta']),
+
+    config: Template.config,
+    name: Template.name,
+    version: Template.version,
+    description: Template.description,
+    maintainer: Template.maintainer,
+    scmUri
+};
+
+module.exports = {
+    /**
+     * All the available properties of Template
+     *
+     * @property base
+     * @type {Joi}
+     */
+    base: Joi.object(MODEL).label('Template'),
+
+    /**
+     * Properties for template that will come back during a GET request
+     *
+     * @property get
+     * @type {Joi}
+     */
+    get: Joi.object(mutate(MODEL, [
+        'id', 'labels', 'config', 'name', 'version', 'description', 'maintainer', 'scmUri'
+    ], [])).label('Get Template'),
+
+    /**
+     * Properties for template that will be passed during a CREATE request
+     *
+     * @property create
+     * @type {Joi}
+     */
+    create: Joi.object(mutate(MODEL, [
+        'config', 'name', 'version', 'description', 'maintainer'
+    ], ['labels'])).label('Create Template'),
+
+    /**
+     * Properties for template that will be passed during a UPDATE requeste
+     *
+     * @property update
+     * @type {Joi}
+     */
+    update: Joi.object(mutate(MODEL, [], ['labels']))
+        .label('Update Template'),
+
+    /**
+     * List of fields that determine a unique row
+     *
+     * @property keys
+     * @type {Array}
+     */
+    keys: ['name', 'version'],
+
+    /**
+     * List of all fields in the model
+     * @property allKeys
+     * @type {Array}
+     */
+    allKeys: Object.keys(MODEL),
+
+    /**
+     * List of indexes to create in the datastore
+     *
+     * @property indexes
+     * @type {Array}
+     */
+    indexes: ['name'],
```

Comment:
Question: with RDS, are indices defined by the `data-schema`? Or is the schema defined separately?

The reason why I ask is because the interface for declaring a template is `template_name@version`. Given this format, I expect `version` to be queried a lot as well as the `name`.